audio/pcm_decode.c

```diff
@@ -356,17 +363,45 @@ static bool pcm_parsewav(FAR struct pcm_decode_s *priv, uint8_t *data)
   localwav.fmt.align     = pcm_leuint16(wav->fmt.align);
   localwav.fmt.bpsamp    = pcm_leuint16(wav->fmt.bpsamp);
 
-  localwav.data.chunkid  = pcm_leuint32(wav->data.chunkid);
-  localwav.data.chunklen = pcm_leuint32(wav->data.chunklen);
+  /* Find the data chunk */
+
+  dchunk = &wav->data;
+
+  while (len < ret)
```

Comment:
should the condition is ret+ 8 <= len?

Reply:
or change to:
```
for (; ; )
  {
      localwav.data.chunkid  = pcm_leuint32(dchunk->chunkid);
      localwav.data.chunklen = pcm_leuint32(dchunk->chunklen);

      if (localwav.data.chunkid == WAV_DATA_CHUNKID)
        {
          break;
        }

      /* Not data chunk. Skip it. */

      ret += localwav.data.chunklen + 8;
     if (ret +8 >= len)
       {
         return -EINVAL;
       }

      dchunk = (FAR const struct wav_datachunk_s *)
               ((uintptr_t)dchunk + localwav.data.chunklen + 8);
  }
```
And remove line 387-392

Reply:
The condition `len < ret` is ok  because:
- The initial value of `ret` (= `sizeof(struct wav_header_s)`) already contains the size of `struct wav_datachunk_s`.
  In the case of no extra chunk, we can just compare with `ret` and the length of the incoming data `len`. 
- The case of `len == ret` is invalid because it means the incoming data only have wav header and there is no data body to be transferred.

But, your `for` loop seems better because we don't need to compare the values twice. I'd like to modify the loop to yours except the error condition check.

Reply:
BTW, I found another problems:

- The size of the 'LIST' chunk seems not always word-aligned.  If not word-aligned 'LIST' chunk exists before 'data' chunk, getting the chunk ID and length may fails in the architectures which don't support unaligned word access such as Cortex-M0+.
- pcm_leuint32() implementation seems wrong. I think it doesn't work with big-endian architectures.

I'll also fix these.

Reply:
Sure.